src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs

```diff
@@ -0,0 +1,123 @@
+                        package,
+                        nuGetFramework,
+                        sourceCacheContext,
+                        logger,
+                        cancellationToken);
+
+                    if (packageDependencyInfo != null)
+                    {
+                        results.Add(packageDependencyInfo);
+                    }
+                    else if (includeUnresolved)
+                    {
+                        results.Add(new PackageDependencyInfo(package, null));
+                    }
+                }
+
+                return results;
+            }
+            catch (NuGetProtocolException)
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Given a list of <see cref="PackageDependencyInfo"/> and a collection of <see cref="IPackageWithDependants"/> populates the latter collection with dependency info based on the <paramref name="packageDependencyInfos"/>.
+        /// This method assumes that there is a 1-to-1 relationship between <paramref name="packageDependencyInfos"/> and <paramref name="packagesWithDependants"/>. 
+        /// </summary>
+        /// <remarks>This method will not handle null <paramref name="packageDependencyInfos"/> and <paramref name="packagesWithDependants"/></remarks>
+        public static void PopulateDependants(IList<PackageDependencyInfo> packageDependencyInfos, IEnumerable<IPackageWithDependants> packagesWithDependants)
+        {
+            foreach (var packageDependencyInfo in packageDependencyInfos)
+            {
+                foreach (var dependency in packageDependencyInfo.Dependencies)
+                {
+                    var matchingDependencyItem = packagesWithDependants
+                        .FirstOrDefault(d => (d.Identity.Id == dependency.Id) && (d.Identity.Version == dependency.VersionRange.MinVersion));
+                    if (matchingDependencyItem != null)
+                    {
+                        matchingDependencyItem.DependingPackages.Add(packageDependencyInfo);
+                    }
+                }
+            }
+        }
+        /// <summary>
+        /// Given <paramref name="packageDependencyInfos"/> generates a collection of <see cref="IPackageWithDependants"/> with the dependants populated correctly.
+        /// </summary>
+        /// <returns>A collection of <see cref="IPackageWithDependants"/></returns>
+        public static IEnumerable<IPackageWithDependants> GetPackagesWithDependants(IList<PackageDependencyInfo> packageDependencyInfos)
+        {
+            var packageWithDependants = packageDependencyInfos.Select(e => new PackageWithDependants(e)).ToList<IPackageWithDependants>();
+            PopulateDependants(packageDependencyInfos, packageWithDependants);
+            return packageWithDependants;
+        }
+
+        private class PackageWithDependants : IPackageWithDependants
```

Comment:
This class is also declared in the `PackageGraphAnalysisUtilitiesTests.cs`. Can we avoid duplication ?

Reply:
I can make it internal. 
I don't really want to make this external as it's an extremely primitive and even unusable model :)